### PR TITLE
Add data transform fn

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/6823

### CODE REVIEW

File: paddle/framework/data_transform.cc [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace framework {
+
+static DataTransformFnMap* data_transform_map = nullptr;
+
+DataTransformFnMap& DataTransformFnMap::Instance() {
+  if (data_transform_map == nullptr) {
+    data_transform_map = new DataTransformFnMap();
+  }
+  return *data_transform_map;
+}
```
Comment (QiJune) [inline paddle/framework/data_transform.cc]: Following is a thread-safe singleton:
```
DataTransformFnMap& DataTransformFnMap::Instance() {
  static DataTransformFnMap inst;
  return inst;  
}
```
refer to https://stackoverflow.com/questions/11711920/how-to-implement-multithread-safe-singleton-in-c11-without-using-mutex

File: paddle/framework/data_transform.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+#include "paddle/framework/op_kernel_type.h"
+#include "paddle/framework/tensor.h"
+#include "paddle/platform/device_context.h"
+#include "paddle/platform/macros.h"
+
+namespace paddle {
+namespace framework {
+
+using DataTransformFN =
```
Comment (reyoung) [inline paddle/framework/data_transform.h]: DataTransform should not only use `Tensor` as its argument type.

How about `SelectedRows`? `LoDTensor`? `vector<Scope>`? It should take Variable as its argument.

Comment (jacquesqiao) [reply paddle/framework/data_transform.h]: ok

Comment (jacquesqiao) [reply paddle/framework/data_transform.h]: done

File: paddle/framework/data_transform.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/framework/data_transform.cc]: done

File: paddle/framework/data_transform.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                       const Tensor& in, Tensor* out)>;
+using KernelTypePair = std::pair<OpKernelType, OpKernelType>;
+
+struct KernelTypePairHash {
+  size_t operator()(const KernelTypePair& kernel_pair) const {
+    OpKernelType::Hash kernel_type_hasher;
+    size_t left_hasher = kernel_type_hasher(kernel_pair.first) << 1;
+    size_t right_hasher = kernel_type_hasher(kernel_pair.second);
+    std::hash<size_t> hasher;
+    return hasher(left_hasher + right_hasher);
```
Comment (QiJune) [inline paddle/framework/data_transform.h]: A little concern on the correctness of hash calculation.

Comment (jacquesqiao) [reply paddle/framework/data_transform.h]: ok, I find a better way to do hash

Comment (jacquesqiao) [reply paddle/framework/data_transform.h]: thanks, done

File: paddle/framework/data_transform.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace paddle {
+namespace framework {
+
+using DataTransformFN =
+    std::function<void(const std::vector<platform::DeviceContext*> ctx,
+                       const Variable& in, Variable* out)>;
+using KernelTypePair = std::pair<OpKernelType, OpKernelType>;
+
+static void hash_combine(std::size_t& seed, const OpKernelType& t) {
```
Comment (QiJune) [inline paddle/framework/data_transform.h]: Code style is:
```
template <class T>
inline void HashCombine(const T& v, std::size_t* seed) {
  std::hash<T> hasher;
  *seed ^= hasher(v) + 0x9e3779b9 + (*seed << 6) + (*seed >> 2);
}
``` 
We can fix it later.

**Pull request merged**